### PR TITLE
Remove image and daemon dependencies from builder

### PR DESCRIPTION
**Pull request by tiborvass**
This is part of the ongoing work to move builder in a standalone repository.

Ping @calavera @duglin @anusha-ragunathan

### CODE REVIEW

File: api/server/router/container/backend.go [hunk 1]
```diff
@@ -32,7 +32,7 @@ type copyBackend interface {
 
 // stateBackend includes functions to implement to provide container state lifecycle functionality.
 type stateBackend interface {
-	ContainerCreate(params *daemon.ContainerCreateConfig) (types.ContainerCreateResponse, error)
+	ContainerCreate(name string, config *runconfig.Config, hostConfig *runconfig.HostConfig, adjustCPUShares bool) (types.ContainerCreateResponse, error)
```
Comment (tiborvass) [inline api/server/router/container/backend.go]: @calavera this is different from what you did in #18519, do you prefer to move things to `api/types` here like you did it in that other PR?

Comment (calavera) [inline api/server/router/container/backend.go]: I'm okay with leaving it like this. My change looked more straight forward because it only change the package name, but we don't really have guidelines about it, so :+1: to this. :ok_hand:

File: builder/dockerfile/internals.go [hunk 2]
```diff
@@ -189,7 +187,7 @@ func (b *Builder) runContextCommand(args []string, allowRemote bool, allowLocalD
 		return nil
 	}
 
-	container, err := b.docker.ContainerCreate(&daemon.ContainerCreateConfig{Config: b.runConfig})
+	container, err := b.docker.ContainerCreate("", b.runConfig, nil, false)
```
Comment (cpuguy83) [inline builder/dockerfile/internals.go]: I'll be honest, I like the config struct better, but this is ok.

File: api/types/configs.go [hunk 3]
```diff
@@ -8,6 +8,14 @@ import (
 	"github.com/docker/docker/runconfig"
 )
 
+// ContainerCreateConfig is the parameter set to ContainerCreate()
+type ContainerCreateConfig struct {
```
Comment (MHBauer) [inline api/types/configs.go]: I like this part, for sure.

File: daemon/create.go [hunk 4]
```diff
@@ -37,19 +29,19 @@ func (daemon *Daemon) ContainerCreate(params *ContainerCreateConfig) (types.Cont
...
-		return types.ContainerCreateResponse{ID: "", Warnings: warnings}, daemon.imageNotExistToErrcode(err)
+		return types.ContainerCreateResponse{Warnings: warnings}, daemon.imageNotExistToErrcode(err)
 	}
 
 	return types.ContainerCreateResponse{ID: container.ID, Warnings: warnings}, nil
 }
 
 // Create creates a new container from the given configuration with a given name.
-func (daemon *Daemon) create(params *ContainerCreateConfig) (retC *container.Container, retErr error) {
+func (daemon *Daemon) create(params types.ContainerCreateConfig) (reC *container.Container, retErr error) {
```
Comment (MHBauer) [inline daemon/create.go]: Can we get rid of the named return values, and create a `var retErr` below?

File: daemon/daemonbuilder/builder.go [hunk 5]
```diff
@@ -84,16 +92,14 @@ func (d Docker) ContainerUpdateCmd(cID string, cmd []string) error {
...
-	//d.Daemon.Graph().Retain(sessionID, imgID)
-}
-
-// Release releases a list of images that were retained for the time of a build.
-func (d Docker) Release(sessionID string, activeImages []string) {
-	// FIXME: This will be solved with tags in client-side builder
-	//d.Daemon.Graph().Release(sessionID, activeImages...)
+// ContainerAttach attaches streams to the container cID. If stream is true, it streams the output.
+func (d Docker) ContainerAttach(cID string, stdin io.ReadCloser, stdout, stderr io.Writer, stream bool) error {
+	return d.Daemon.ContainerWsAttachWithLogs(cID, &daemon.ContainerWsAttachWithLogsConfig{
```
Comment (MHBauer) [inline daemon/daemonbuilder/builder.go]: I feel like we can move this config to `api/types/configs.go`, as well as `ContainerAttachWithLogsConfig`, but that might be better done in a different PR.

File: builder/builder.go [hunk 6]
```diff
@@ -112,13 +110,13 @@ type Backend interface {
 	// TODO: use digest reference instead of name
 
 	// GetImage looks up a Docker image referenced by `name`.
-	GetImage(name string) (*image.Image, error)
+	GetImage(name string) (Image, error)
 	// Pull tells Docker to pull image referenced by `name`.
-	Pull(name string) (*image.Image, error)
-	// ContainerWsAttachWithLogs attaches to container.
-	ContainerWsAttachWithLogs(name string, cfg *daemon.ContainerWsAttachWithLogsConfig) error
```
Comment (runcom) [inline builder/builder.go]: nit picking about having a struct for the new `ContainerAttach` to be somehow consistent with other methods - `ContainerRm` has a struct for just 3 arguments for instance

Comment (tiborvass) [inline builder/builder.go]: I wonder why `ContainerRm` has a struct for just 3 arguments :)

Comment (tiborvass) [inline builder/builder.go]: If not having a struct for attach really is a blocker for people, I'm fine addressing it.

Comment (runcom) [inline builder/builder.go]: nah it isn't (just wanted to note that we have structs for 3 args and plain args for 4 here)

**Comment by runcom** [comment]
LGTM

**Comment by vdemeester** [comment]
LGTM 🐹

File: daemon/create.go [hunk 7]
```diff
@@ -37,24 +29,25 @@ func (daemon *Daemon) ContainerCreate(params *ContainerCreateConfig) (types.Cont
...
-		return types.ContainerCreateResponse{ID: "", Warnings: warnings}, daemon.imageNotExistToErrcode(err)
+		return types.ContainerCreateResponse{Warnings: warnings}, daemon.imageNotExistToErrcode(err)
 	}
 
 	return types.ContainerCreateResponse{ID: container.ID, Warnings: warnings}, nil
 }
 
 // Create creates a new container from the given configuration with a given name.
-func (daemon *Daemon) create(params *ContainerCreateConfig) (retC *container.Container, retErr error) {
+func (daemon *Daemon) create(params types.ContainerCreateConfig) (*container.Container, error) {
```
Comment (tonistiigi) [inline daemon/create.go]: @tiborvass This doesn't look right. Cleanup is unhandled now because returns don't set `retErr`.

Comment (dmcgowan) [inline daemon/create.go]: I had to revert this part of the change to get the cleanup test passing for #18834. A failed run was leaving around a container because it was unable to trigger the removal.
